src/ApacheOrcDotNet.OptimizedReader/OrcReader.cs

```diff
@@ -35,93 +40,452 @@ public OrcReader(OrcReaderConfiguration configuration, IByteRangeProvider byteRa
-                (int)_fileTail.PostScript.CompressionBlockSize,
-                sequence => SpanStripeFooter.ReadStreamDetails(sequence, ColumnDetails, (long)stripe.Offset)
-            );
+            columnBuffer.Fill();
 
-            return result;
+            if (NumValues == 0 || NumValues > columnBuffer.Values.Length)
+                NumValues = columnBuffer.Values.Length;
         }
 
-        SpanFileTail ReadFileTail()
+        public ColumnStatistics GetFileColumnStatistics(int columnId)
+            => _fileTail.Footer.Statistics[columnId];
+
+        public ColumnStatistics GetStripeColumnStatistics(int columnId, int stripeId)
+            => _fileTail.Metadata.StripeStats[stripeId].ColStats[columnId];
+
+        public RowIndex GetColumnRowIndex(int columnId, int stripeId)
+        {
+            var key = (columnId, stripeId);
+
+            return _rowGroupIndexes.GetOrAdd(key, key =>
+            {
+                var streamDetails = GetStripeStreams(stripeId);
+                var rowIndexStream = streamDetails.Where(s =>
+                    s.StreamKind == StreamKind.RowIndex
+                    && s.ColumnId == columnId
+                ).Single();
+
+                var decompressedData = _byteRangeProvider.DecompressByteRange(
+                     rowIndexStream.FileOffset,
+                     rowIndexStream.Length,
+                     _fileTail.PostScript.Compression,
+                     (int)_fileTail.PostScript.CompressionBlockSize
+                 );
+
+                return Serializer.Deserialize<RowIndex>(decompressedData.Sequence);
+            });
+        }
+
+        private SpanFileTail ReadFileTail()
         {
             int lengthToReadFromEnd = _configuration.OptimisticFileTailReadLength;
             while (true)
             {
-                var buffer = ArrayPool<byte>.Shared.Rent(lengthToReadFromEnd);
-                var bufferSpan = buffer.AsSpan()[..lengthToReadFromEnd];
-                _byteRangeProvider.GetRangeFromEnd(bufferSpan, lengthToReadFromEnd);
-                var success = SpanFileTail.TryRead(bufferSpan, out var fileTail, out var additionalBytesRequired);
-                ArrayPool<byte>.Shared.Return(buffer);
+                var fileTailBufferRaw = ArrayPool<byte>.Shared.Rent(lengthToReadFromEnd);
+                var fileTailBuffer = fileTailBufferRaw.AsSpan()[..lengthToReadFromEnd];
+
+                _byteRangeProvider.GetRangeFromEnd(fileTailBuffer, lengthToReadFromEnd);
+
+                var success = SpanFileTail.TryRead(fileTailBuffer, out var fileTail, out var additionalBytesRequired);
+                ArrayPool<byte>.Shared.Return(fileTailBufferRaw);
 
                 if (success)
                     return fileTail;
 
                 lengthToReadFromEnd += additionalBytesRequired;
             }
         }
+
+        private IEnumerable<StreamDetail> GetStripeStreams(int stripeId)
+        {
+            if (!_stripeStreams.ContainsKey(stripeId))
```

Comment:
Why doesn't this need to be concurrent?

Reply:
It was kept like this because it is only called from within blocks that are on a thread safe context (`GetColumnRowIndex` and `GetColumnDataStreams`).